main/src/library/Applicative.flix

```diff
@@ -0,0 +1,83 @@
+/*
+ *  Copyright 2021 Felix Wiemuth
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+///
+/// A type class for functors that support application, i.e. allow to:
+/// - Make an applicative value out of a normal value (embed it into a default context), e.g. embed `5` into `Some(5)`.
+/// - Apply a function-type applicative to a matching argument-type applicative, resulting in an applicative of
+/// the function's result type. The meaning of this application is defined by the respective instance.
+/// Conceptually this can be understood as applying functions "contained" in the first applicative to arguments
+/// in the second applicative, where the possible quantity of functions/arguments depends on the type `m`.
+/// For example, an `Option[a -> b]` can be `None`, or contain a function of type `a -> b`, and only in the
+/// latter case a function is applied. A `List[a -> b]` contains a list of functions, which are all to be
+/// applied to all arguments in the arguments list.
+///
+/// A minimal implementation must define `point` and at least one of `ap` and `liftA2`. If it defines both,
+/// they must be equivalent to their default definitions:
+///     `ap(f: m[a -> b & e], x: m[a]) : m[b] & e = liftA2(identity, f, x)`
+///     `liftA2(f: a -> b -> c & e, x: m[a], y: m[b]): m[c] & e = ap(Functor.map(f, x), y)`
+///
+class Applicative[m : Type -> Type] with Functor[m] {
+    ///
+    /// Puts `x` into a default context.
+    ///
+    pub def point(x: a) : m[a]
+
+    ///
+    /// Apply the function-type applicative `f` to the argument-type applicative `x`.
+    ///
+    pub def ap(f: m[a -> b & e], x: m[a]) : m[b] & e = Applicative.liftA2(identity, f, x)
+
+    ///
+    /// Lift a binary functions to work on `Applicative`s.
+    /// Instances can define more efficient implementations than the default implementation
+    /// (which is `Applicative.ap(Functor.map(f, x), y)`).
+    ///
+    pub def liftA2(f: a -> b -> c & e, x: m[a], y: m[b]): m[c] & e = Applicative.ap(Functor.map(f, x), y)
```

Comment:
What about liftAX? X> 2

Reply:
> Whats the plan for liftX?

Yes I guess it could be useful to have `liftAX` for X > 2, especially as this allows for more efficient implementations than the default implementation, which would be e.g. `liftA3(f, x1, x2, x3) = ap(liftA2(f, x1, x2), x3)` (or without indirections `ap(ap(map(f, x1), x2), x3)`).

For example, for the list instance, an implementation of `liftAX` which directly applies the function `f` to all combinations of the X arguments would be more efficient than creating a list of partially applied functions for each nested `ap` call (until reaching `liftA2` which might have the more efficient implementation).

For the `Option` instance, however, I guess the default implementation is the best one can do (except maybe for saving a few function calls).

One could also define lift for X=1, i.e. `liftA1(f: a -> b & e, x : m[a]) : m[b] = ap(point(f), x)`. But just as we have a law for `liftA2` requiring that `liftA2(f, x, y) == ap(map(f, x), y)`, we would require `liftA1(f,x) == map(f,x)`, so there is maybe no point in adding `liftA1`.

> Also the difference in naming between lift and liftA was not entirely clear to me.

I would say to distinguish applicative lift (`liftA`) from the corresponding monadic lift (`liftM`). Probably one eventually wants both in Prelude without namespacing prefix, and in the type classes where the instances are defined they of course also need to be distinguished.

It might still make sense to define `liftX` for e.g. `Option`, representing lifting of a function to work on `Option`s - here `liftX` would serve as implementation for both `liftAX` and `liftMX`.

Reply:
OK, so the actionable items are:

- [ ] Define liftXA in the TC. 
- [ ] Rename `Option.liftX` to `Option.liftXA` 

Does you agree? Do you want to do it in a separate PR?

Reply:
I would complete all the outstanding small changes in this PR.

I'm not sure about renaming e.g. `Option.liftX` to `Option.liftAX`. The argument was that this "lift" is also somewhat independent of Applicative or Monad. So I actually like that it is neutrally called `liftX`, while `liftAX` and `liftMX` will be defined based on that.

Reply:
OK and OK